crates/bevy_ecs/src/bundle.rs

```diff
@@ -75,26 +77,106 @@ use std::{any::TypeId, collections::HashMap};
     /// [`Component`]s
-    unsafe fn from_components(func: impl FnMut() -> *mut u8) -> Self
-    where
-        Self: Sized;
+    unsafe fn from_components(func: impl FnMut() -> *mut u8) -> Self;
+
+    /// Gets this [`Bundle`]'s component ids, in the order of this bundle's [`Component`]s
+    fn component_ids(components: &mut Components, storages: &mut Storages) -> Vec<ComponentId>;
 
     /// Calls `func` on each value, in the order of this bundle's [`Component`]s. This will
     /// [`std::mem::forget`] the bundle fields, so callers are responsible for dropping the fields
     /// if that is desirable.
     fn get_components(self, func: impl FnMut(*mut u8));
 }
 
+mod sealed {
+    use super::{ApplicableBundle, Bundle};
+
+    pub trait SealedApplicableBundle {}
+    impl<T> SealedApplicableBundle for T where T: Bundle {}
+    impl SealedApplicableBundle for Box<dyn ApplicableBundle> {}
+}
+
+/// A bundle which can be applied to a world
+///
+/// # Safety
+/// The bundle returned from `init_bundle_info` is the same as used for `get_component_box`.
+///
+/// This trait is sealed and cannot be implemented outside of `bevy_ecs`
+///
+/// However, it is implemented for every type which implements [`Bundle`]
+pub unsafe trait ApplicableBundle:
+    Send + Sync + 'static + sealed::SealedApplicableBundle
+{
+    fn init_bundle_info<'a>(
+        &self,
+        bundles: &'a mut Bundles,
+        components: &mut Components,
+        storages: &mut Storages,
+    ) -> &'a BundleInfo;
+    // Same requirements as [`Bundle::get_components`]
+    fn get_components_box(self: Box<Self>, func: &mut dyn FnMut(*mut u8));
+    // Same requirements as [`Bundle::get_components`]
+    fn get_components_self(self, func: impl FnMut(*mut u8))
+    where
+        Self: Sized;
+}
+
+unsafe impl<T> ApplicableBundle for T
+where
+    T: Bundle,
+{
+    fn init_bundle_info<'a>(
+        &self,
+        bundles: &'a mut Bundles,
+        components: &mut Components,
+        storages: &mut Storages,
+    ) -> &'a BundleInfo {
+        bundles.init_info::<Self>(components, storages)
+    }
+
+    fn get_components_box(self: Box<Self>, func: &mut dyn FnMut(*mut u8)) {
+        Self::get_components(*self, func)
+    }
+
+    fn get_components_self(self, func: impl FnMut(*mut u8))
+    where
+        Self: Sized,
+    {
+        Self::get_components(self, func)
+    }
+}
+
+unsafe impl ApplicableBundle for Box<dyn ApplicableBundle> {
+    fn init_bundle_info<'a>(
+        &self,
+        bundles: &'a mut Bundles,
+        components: &mut Components,
+        storages: &mut Storages,
+    ) -> &'a BundleInfo {
+        <dyn ApplicableBundle as ApplicableBundle>::init_bundle_info(
+            self, bundles, components, storages,
+        )
+    }
+
+    fn get_components_box(self: Box<Self>, func: &mut dyn FnMut(*mut u8)) {
+        <dyn ApplicableBundle as ApplicableBundle>::get_components_box(self, func)
```

Comment:
Mildly confused: why doesn't this just infinitely recurse? It clearly seems to *work*, but I don't understand why.

Reply:
It's possible that this does infinitely recurse; it's untested, and only useful in the case where you have `Box<dyn ApplicableBundle>`, where the inner `ApplicableBundle` is another `Box<dyn ApplicableBundle>`

This is the implementation on `Box<dyn ApplicableBundle>`; it's calling the impl of `dyn ApplicableBundle`, which has a self type of `Box<dyn ApplicableBundle>`.

Reply:
Aha, I see! I'd misread the scope of this.